### PR TITLE
Hash file paths and lengths when hashing a directory

### LINKED ISSUE #5254: `Sha` class causes hash collisions for different directory trees due to lack of delimiter and absence of file paths in hashing
_Requested by **Yegor Bugayenko** · [Slack thread](https://eolang-org.slack.com/archives/C08932P96NM/p1782286325560679)_

### Summary

`org.eolang.maven.Sha` is documented to compute a SHA-256 hash of a directory by walking "all files sorted by path and hashes their combined contents". When hashing a directory, it feeds the raw bytes of every file into a single `MessageDigest` back-to-back with **no per-file length delimiter and without hashing the file paths**. As a result, two clearly different directory trees can produce identical hashes — a genuine collision that breaks the contract the class advertises.

### Location

https://github.com/objectionary/eo/blob/master/eo-maven-plugin/src/main/java/org/eolang/maven/Sha.java#L54-L76

(permalink pinned at HEAD `382ead173c708672b4e31dac024308402695d1df`)

```java
private static String hash(final Path path) throws IOException, NoSuchAlgorithmException {
    final MessageDigest digest = MessageDigest.getInstance("SHA-256");
    try (Stream<Path> walk = Files.walk(path)) {
        walk.filter(Files::isRegularFile)
            .sorted(Comparator.comparing(Path::toString)).forEach(
                file -> {
                    try (InputStream input = Files.newInputStream(file)) {
                        final byte[] buffer = new byte[8192];
                        int read = input.read(buffer);
                        while (read != -1) {
                            digest.update(buffer, 0, read);   // <-- only raw content, no boundary, no path
                            read = input.read(buffer);
                        }
                    } catch (final IOException ex) {
                        throw new IllegalStateException(
                            String.format("Failed to read '%s'", file), ex
                        );
                    }
                }
            );
    }
    return Base64.getEncoder().encodeToString(digest.digest());
}
```

The class-level Javadoc explicitly promises directory support:

```java
/**
 * SHA-256 hash of a file or directory.
 * For a directory, walks all files sorted by path and hashes their combined contents.
 * @since 0.62.0
 */
```

### Why it's wrong

Because the files are sorted by path and then only their **contents** are streamed into one digest, the hash depends solely on the concatenation of all bytes — the boundaries between files and the file names themselves are invisible to the digest. Any two directory trees whose sorted-by-path content concatenation is byte-identical hash to the same value, even though the trees are different.

Concrete, reproduced collision using the exact logic of `Sha.hash`:

| Directory | Files (sorted by path) | Concatenated bytes | SHA-256 (Base64) |
|-----------|------------------------|--------------------|------------------|
| A | `file1="ab"`, `file2="c"` | `abc` | `ungWv48Bz+pBQUDeXa4iI7ADYaOWF3qctBD/YfIAFa0=` |
| B | `file1="a"`,  `file2="bc"` | `abc` | `ungWv48Bz+pBQUDeXa4iI7ADYaOWF3qctBD/YfIAFa0=` |

Both yield `ungWv48Bz+pBQUDeXa4iI7ADYaOWF3qctBD/YfIAFa0=` (the SHA-256 of `"abc"`), so `Sha(A).equals(Sha(B))` despite A and B being distinct directories. Renaming a file (same content, different name) likewise leaves the hash unchanged, since paths are used only for ordering, never hashed.

Notably, `Cache.dirSha` (the production path that actually hashes directories) does **not** reuse `Sha`'s directory branch — it independently hashes each file to a fixed-length per-file SHA string and concatenates those. That separate reimplementation exists precisely because raw-content concatenation is unsafe, which corroborates that `Sha`'s own directory handling is broken.

There is no `ShaTest` in `eo-maven-plugin/src/test`, so this broken edge case is currently untested.

### Expected behavior

Distinct directory trees (different file boundaries, different file names, or different layouts) should produce distinct hashes. The hash should depend on both each file's path (relative to the root) and its content, in an unambiguous, length-delimited way — matching the "walks all files sorted by path" contract.

### Suggested fix

Make file boundaries and paths part of the digest. For example, for each file (in sorted order) feed a delimiter derived from the relative path and the content length before the content:

```java
final Path rel = path.relativize(file);
digest.update(rel.toString().getBytes(StandardCharsets.UTF_8));
digest.update((byte) 0);                                  // unambiguous separator
final long size = Files.size(file);
digest.update(ByteBuffer.allocate(Long.BYTES).putLong(size).array());
// ... then stream the file content as today ...
```

Equivalently, hash each file individually (like `Cache.dirSha` already does) and combine the fixed-length per-file digests together with the relative path. Adding a `ShaTest` that asserts the A/B trees above produce different hashes would lock in the fix.

### PR DESCRIPTION
**Pull request by Thayorns**
`Sha` streamed the raw bytes of every file of a directory into one digest, back to back. Nothing in the digest said where one file ended and the next began, and the file names were used only for sorting, never hashed. Two trees whose sorted contents concatenate to the same bytes therefore got the same hash: `{a.txt="ab", b.txt="c"}` and `{a.txt="a", b.txt="bc"}` both came out as the SHA-256 of `abc`, and renaming a file left the hash untouched.

Now every file of a directory is preceded in the digest by its path relative to the root and followed by the number of bytes read from it, both NUL-terminated, so the boundaries and the names are part of what gets hashed. The length comes from the read loop itself rather than from a second `Files.size` call, so it always describes the bytes that were really hashed even if the file changes size underneath us. Relative paths go into the digest, and serve as the sort key, with forward slashes on every platform, so the same tree hashes alike on Windows and on Linux. A lone regular file is left alone and still hashes to the plain SHA-256 of its bytes, which is what `CacheTest` pins down and what the `.sha256` markers already on disk contain.

Only files speak into the digest, so a directory holding none of them stays invisible to the hash, the way Git also sees it. That is now written down in the class docblock. Framing bare directory paths too would trade one collision for another: `{1/, 234/, 3/}` as empty directories and a single file `1` holding `234` both come out as `\0` `1\0` `234\0` `3\0`.

`ShaTest` covers the three collisions above, a file moved to a differently named subdirectory, and the plain hash of a lone file, and checks that two identical trees sitting in different directories still agree, so the relative path never leaks the absolute one into the hash.

`Cache.dirSha` still hashes directories on its own instead of going through `Sha`, which is why the bug never showed up in production. It cannot simply delegate yet, because it also skips files that `Sha` has no way of skipping, so it carries a puzzle now.

Closes #5254

### CODE REVIEW

**Comment by Thayorns** [comment]
@yegor256 take a look please, happy to clarify anything here.

File: eo-maven-plugin/src/main/java/org/eolang/maven/Sha.java [hunk 1]
```diff
@@ -16,7 +17,8 @@
 
 /**
  * SHA-256 hash of a file or directory.
- * For a directory, walks all files sorted by path and hashes their combined contents.
+ * For a directory, hashes every file sorted by path, framed by its relative path and length,
```
Comment (morphqdd) [inline eo-maven-plugin/src/main/java/org/eolang/maven/Sha.java]: The PR description says Cache.dirSha still hashes directories on its own instead of going through this class, which is why the old collision never showed up in production, and that folding it in is worth doing later. Nothing in the code marks that as outstanding work, only the PR body. A puzzle here (or on Cache.dirSha itself) would let 0pdd track it as a real issue instead of it depending on someone rereading this PR's description.

File: eo-maven-plugin/src/main/java/org/eolang/maven/Sha.java [hunk 2]
```diff
@@ -37,27 +39,34 @@ final class Sha {
...
+        final boolean tree = Files.isDirectory(this.path);
+        try (Stream<Path> walk = Files.walk(this.path)) {
             walk.filter(Files::isRegularFile)
                 .sorted(Comparator.comparing(Path::toString)).forEach(
                     file -> {
                         try (InputStream input = Files.newInputStream(file)) {
+                            if (tree) {
+                                digest.update(
+                                    String.format(
+                                        "%s\0%d\0", this.path.relativize(file), Files.size(file)
```
Comment (morphqdd) [inline eo-maven-plugin/src/main/java/org/eolang/maven/Sha.java]: Files.size(file) is a separate stat call from the read loop just below that actually streams and hashes the file's bytes. If the file changes size between the two (a build touching its own inputs, a concurrent writer), the length framed into the digest would not match what was actually hashed, undermining the exact boundary guarantee this fix is meant to add. Using the byte count accumulated by the read loop itself would tie the framed length to what was really hashed.

File: eo-maven-plugin/src/main/java/org/eolang/maven/Sha.java [hunk 3]
```diff
@@ -37,27 +39,34 @@ final class Sha {
...
      * @throws NoSuchAlgorithmException If SHA-256 is unavailable
      */
-    private static String hash(final Path path) throws IOException, NoSuchAlgorithmException {
+    private String hash() throws IOException, NoSuchAlgorithmException {
         final MessageDigest digest = MessageDigest.getInstance("SHA-256");
-        try (Stream<Path> walk = Files.walk(path)) {
+        final boolean tree = Files.isDirectory(this.path);
+        try (Stream<Path> walk = Files.walk(this.path)) {
             walk.filter(Files::isRegularFile)
                 .sorted(Comparator.comparing(Path::toString)).forEach(
```
Comment (morphqdd) [inline eo-maven-plugin/src/main/java/org/eolang/maven/Sha.java]: Both the sort key here and the relative path framed into the digest go through Path::toString(), which uses backslashes on Windows and forward slashes elsewhere. The same directory tree would sort differently and hash to a different digest depending on the OS it is hashed on, which matters if this hash is ever used as a cache key shared or compared across platforms.

File: eo-maven-plugin/src/main/java/org/eolang/maven/Sha.java [hunk 4]
```diff
@@ -37,27 +39,34 @@ final class Sha {
...
      * @throws IOException If reading fails
      * @throws NoSuchAlgorithmException If SHA-256 is unavailable
      */
-    private static String hash(final Path path) throws IOException, NoSuchAlgorithmException {
+    private String hash() throws IOException, NoSuchAlgorithmException {
         final MessageDigest digest = MessageDigest.getInstance("SHA-256");
-        try (Stream<Path> walk = Files.walk(path)) {
+        final boolean tree = Files.isDirectory(this.path);
+        try (Stream<Path> walk = Files.walk(this.path)) {
             walk.filter(Files::isRegularFile)
```
Comment (morphqdd) [inline eo-maven-plugin/src/main/java/org/eolang/maven/Sha.java:61]: Two directories that both contain zero regular files hash the same regardless of their names or nesting, since nothing here contributes anything to the digest for a directory with no files in it. Not one of the collisions this PR sets out to fix, and not mentioned as a remaining gap either.

File: eo-maven-plugin/src/test/java/org/eolang/maven/ShaTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        new Saved(text.substring(7), temp.resolve("second/b.txt")).value();
+        MatcherAssert.assertThat(
+            String.format("hashes of differently split dirs are equal, seed=%d", seed),
+            new Sha(temp.resolve("first")).toString(),
+            Matchers.not(Matchers.equalTo(new Sha(temp.resolve("second")).toString()))
+        );
+    }
+
+    @Test
+    void distinguishesDirsWithRenamedFile(@Mktmp final Path temp) throws IOException {
```
Comment (morphqdd) [inline eo-maven-plugin/src/test/java/org/eolang/maven/ShaTest.java:43]: This only renames a file at the top level. There is no test for the same file name moving to a different nested subdirectory (for example sub1/x.txt vs sub2/x.txt), which exercises the same relative-path framing but at a deeper level than the flat rename here.

**Comment by 0crat** [comment]
@morphqdd Thanks for the review! You've earned +7 points for this: +12 as a basis; -5 for very few (5) comments.  Your running score is [+3765](https://www.eolang.org/zerocracy/objectionary-vitals.html); don't forget to check your [Zerocracy account](https://www.zerocracy.com/) too).

**Comment by Thayorns** [comment]
@yegor256 have a look

**Comment by 0crat** [comment]
@thayorns Thanks for the contribution! You've earned +12 points for this: +16 as a basis; -4 for too many hits-of-code (195 >= 100). Please, [keep](https://www.yegor256.com/2018/03/06/speed-vs-quality.html) them coming. Your running score is [+3935](https://www.eolang.org/zerocracy/objectionary-vitals.html); don't forget to check your [Zerocracy account](https://www.zerocracy.com/) too).
